File: linux/tabletbridge/protocol.py
"""Wire format shared with the Android app (see Protocol.kt).

Every message is: 1 byte type, 4 byte big-endian payload length, payload.
"""
from __future__ import annotations

import json
import socket
import struct
import threading

PORT = 27183
VERSION = 1

HELLO = 0x01        # JSON {"version": 1}
MIC_PCM = 0x10      # tablet -> laptop, s16le 48k mono
SPK_PCM = 0x20      # laptop -> tablet, s16le 48k stereo
FRAME_JPEG = 0x30   # laptop -> tablet
CONTROL = 0x40      # laptop -> tablet, JSON {"mic": bool}

SAMPLE_RATE = 48000
MAX_PAYLOAD = 16 * 1024 * 1024

_HEADER = struct.Struct("!BI")


class Connection:
    """Thread-safe framed socket wrapper."""

    def __init__(self, sock: socket.socket):
        self.sock = sock
        self._send_lock = threading.Lock()
        self._rfile = sock.makefile("rb", buffering=256 * 1024)

    def send(self, msg_type: int, payload: bytes) -> None:
        with self._send_lock:
            self.sock.sendall(_HEADER.pack(msg_type, len(payload)) + payload)

    def send_json(self, msg_type: int, obj: dict) -> None:
        self.send(msg_type, json.dumps(obj).encode())

    def recv(self) -> tuple[int, bytes]:
        header = self._rfile.read(_HEADER.size)
        if len(header) < _HEADER.size:
            raise ConnectionError("connection closed")
        msg_type, length = _HEADER.unpack(header)
        if length > MAX_PAYLOAD:
            raise ConnectionError(f"bad frame length {length}")
        payload = self._rfile.read(length)
        if len(payload) < length:
            raise ConnectionError("connection closed mid-frame")
        return msg_type, payload

    def close(self) -> None:
        try:
            self.sock.shutdown(socket.SHUT_RDWR)
        except OSError:
            pass
        self.sock.close()

File: android/app/src/main/java/com/tabletbridge/Protocol.kt
package com.tabletbridge

import java.io.DataInputStream
import java.io.EOFException
import java.io.OutputStream

/**
 * Wire format (both directions): 1 byte type, 4 byte big-endian payload length, payload.
 * Kept identical to linux/tabletbridge/protocol.py.
 */
object Protocol {
    const val PORT = 27183
    const val VERSION = 1

    const val HELLO: Int = 0x01        // JSON {"version":1}
    const val MIC_PCM: Int = 0x10      // tablet -> laptop, s16le 48k mono
    const val SPK_PCM: Int = 0x20      // laptop -> tablet, s16le 48k stereo
    const val FRAME_JPEG: Int = 0x30   // laptop -> tablet
    const val CONTROL: Int = 0x40      // laptop -> tablet, JSON {"mic":bool}

    const val SAMPLE_RATE = 48000
    const val MAX_PAYLOAD = 16 * 1024 * 1024

    class Message(val type: Int, val payload: ByteArray)

    fun read(input: DataInputStream): Message {
        val type = input.read()
        if (type < 0) throw EOFException()
        val len = input.readInt()
        if (len < 0 || len > MAX_PAYLOAD) throw IllegalStateException("bad frame length $len")
        val payload = ByteArray(len)
        input.readFully(payload)
        return Message(type, payload)
    }

    fun write(out: OutputStream, type: Int, payload: ByteArray, len: Int = payload.size) {
        val header = byteArrayOf(
            type.toByte(),
            (len ushr 24).toByte(), (len ushr 16).toByte(), (len ushr 8).toByte(), len.toByte()
        )
        synchronized(out) {
            out.write(header)
            out.write(payload, 0, len)
            out.flush()
        }
    }
}
